### ISSUE TITLE
mipmaps for pois

### ISSUE THREAD

**Issue by tallytalwar** [tangrams/eraser-map#45]
- [ ] Just do it

Does it make sense for pois to have mipmap filtering?

Refer: https://github.com/tangrams/eraser-map/blob/gh-pages/eraser-map.yaml#L117

**Comment by nvkelso** [comment]
@tallytalwar What alternative should we use?

**Comment by tallytalwar** [comment]
I am not sure, but I believe we do not need any filtering for icons! Karim can add more on it though.

**Comment by karimnaaji** [comment]
Mipmapping basically optimize rendering of icons when they are scaled at a bigger/smaller ratio than their original size. So actually if we don't drastically modify the `size` param on any of the sprite/icons this would save us minimal memory. For eraser-map we would be ok not specifying any filtering and using the default one.

**Comment by nvkelso** [comment]
We do scale the images (in EraserMap), though?

**Comment by karimnaaji** [comment]
Right, just been checking the actual reductions, it would make sense to use this if the reduction is more than half of the original icon size. But personally, I would recommend to use this filtering only when the texture is applied in 3d perspective. I experienced issues when neighboring icons were blurred together during the mipmap process which can cause some annoying artifacts on borders (if we don't space them enough in the icon atlas), so the best is not to use this filtering when providing an atlas.
Once we have the style-sheet working in es I'll test the different cases to see what fits best.

**Comment by nvkelso** [comment]
It’s super easy to space the assets farther apart in the sprite sheet, btw.

(They used to be farther apart, but Patricio needed them piled on top of
each other for a Tangram experiment and we never spaced them back out.)

**Comment by karimnaaji** [comment]
Oh ok I see. If we notice any artifacts,  with this specific filtering we'd try different spacing, for now we can let it as is.

**Comment by karimnaaji** [comment]
To expose more clearly what I was saying yesterday, this is what happens when you provide this filtering which generate those images (where each of them is half the size of the previous one):

Level0:
<img src="https://cloud.githubusercontent.com/assets/7061573/10197837/73333672-6767-11e5-8002-4fa9490c2d00.png" width="256">
Level1:
<img src="https://cloud.githubusercontent.com/assets/7061573/10197839/74fc817a-6767-11e5-884e-7d468f033abf.png" width="256">
Level2:
<img src="https://cloud.githubusercontent.com/assets/7061573/10197841/76564240-6767-11e5-9aac-0dd63302e4fd.png" width="256">
Level3:
<img src="https://cloud.githubusercontent.com/assets/7061573/10197842/776d4638-6767-11e5-9949-0343151300f5.png" width="256">
So clearly, we can't use icons over level1 where information is lost, which is why scaling over half of the size in minification isn't a use case for us, and using the default filtering would be totally fine.

**Comment by nvkelso** [comment]
Cool, thanks for researching. Is the recommended fix to remove line https://github.com/tangrams/eraser-map/blob/gh-pages/eraser-map.yaml#L117?

**Comment by karimnaaji** [comment]
Yep

**Comment by karimnaaji** [comment]
To visualize how it renders in ES, by aggressively scaling down the size icons:

Default:
<img src="https://cloud.githubusercontent.com/assets/7061573/10198680/0f4bcfd0-676b-11e5-9018-f321099e55e1.png" width="210">
Mipmaps:
<img src="https://cloud.githubusercontent.com/assets/7061573/10198671/0262c986-676b-11e5-89ee-fff4c7ae63bc.png" width="210">

**Comment by bcamper** [comment]
@karimnaaji in JS, we default to `linear` filtering (not `nearest` which visually is "no" filtering) if no filtering is specified. I really think `mipmap` is overkill in most of these cases (likely for POIs anyway). It makes sense for bigger textures, like those world texture experiments Patricio was doing, or perhaps larger area decals? Also, in WebGL, mipmapping is not supported for non-power-of-2 textures, so in those cases it falls back to `linear` even if `mipmap` is requested (but ES supports mipmapping for non-power-of-2 I think right?).

In short, we should have support for `filtering` as `mipmap`, `linear`, `nearest` in ES, and default to `linear`. But for this style I think we should just switch to `linear` anyway (it's only `mipmap` because it was copy-pasted from another example).

**Comment by nvkelso** [comment]
Using linear would probably fix a bug I’ve seen where specifying sizes not
clean fractions of the original has no affect.

**Comment by bcamper** [comment]
Um, strange, can you elaborate / have an example? That should not be the case.

**Comment by karimnaaji** [comment]
There's is support for NPOT on desktop, but not on Opengl ES 2.0 (at least for mipmap generation) without using extensions (or on some devices without extensions, but I've seen it working on some devices without doing anything, which depends on the driver), if we want to use mipmap for anything we would need to use power-of-2 textures. Also, there's no fallback on this right now on ES, but I'll follow up and add the behavior.

It's a different but related subject on mipmaps, and I think that @blair1618 would agree with that, since the stylesheet is getting stable on how many/what kind of pattern would be used (mainly grid/dot patterns) we must use textures with mipmap filtering for those (with a default color that we could easily modify in the shader).
One of the main advantage of doing this would be a consistent dot/grid pattern on tilted view, mipmaps are really good for that, pixel shaders just screw the pattern since the covered screen size is hardly known while tilting/rotating which leads to flickering and aliasing on the pattern, but also in rendering (one texture fetch vs a full pixel pattern shader would be much better).

Current status on ES with eraser-map stylesheet:
- grid pattern on rotated/tilted view on a building

<img src="https://cloud.githubusercontent.com/assets/7061573/10405088/b90eb0e2-6ea4-11e5-8d58-af999bb56594.png" width="210">
- dot pattern from far away on tilted view

<img src="https://cloud.githubusercontent.com/assets/7061573/10405120/dfda5f28-6ea4-11e5-91ab-037d83d174b0.png" width="210">

<img src="https://cloud.githubusercontent.com/assets/7061573/10405324/219ec6a0-6ea6-11e5-81ba-25a729fb1296.png" width="210">
